{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "# Домашнее задание №4\n",
    "\n",
    "## Задача №0 (3)\n",
    "Дорешайте задачу про риды и BWT из файла practice4.ipynb. В ответе помимо кода укажите количество ридов первой, второй и третьей категории, а также вкратце опишите ход решения."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysuffixarray.core import SuffixArray\n",
    "from Bio import SeqIO\n",
    "\n",
    "record = SeqIO.read(\"BWT_folder/genome.fa\", \"fasta\")\n",
    "genome_sequence = str(record.seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "class BWTSearcher:\n",
    "    def __init__(self, reference):\n",
    "        # Construct the suffix array\n",
    "        self.sa = SuffixArray(reference)\n",
    "\n",
    "        # Construct the BWT from the suffix array\n",
    "        self.bwt_text = ''.join(reference[i-1] if i != 0 else '$' for i in self.sa.suffix_array())\n",
    "\n",
    "        self.Occ = {ch: [0] * len(self.bwt_text) for ch in \"$ACGT\"}\n",
    "\n",
    "        for i, ch in enumerate(self.bwt_text):\n",
    "            for ch2 in \"ACGT\":\n",
    "                if i != 0:\n",
    "                    self.Occ[ch2][i] = self.Occ[ch2][i - 1]\n",
    "            if ch in self.Occ.keys():\n",
    "                self.Occ[ch][i] += 1\n",
    "\n",
    "        self.count = {}\n",
    "        total = 0\n",
    "        for ch in \"$ACGT\":\n",
    "            self.count[ch] = total\n",
    "            total += self.bwt_text.count(ch)\n",
    "\n",
    "    def bwt_pattern_search(self, pattern):\n",
    "        top = 0\n",
    "        bottom = len(self.bwt_text) - 1\n",
    "        for char in reversed(pattern):\n",
    "            if char in self.Occ:\n",
    "                top = self.count[char] + (0 if top == 0 else self.Occ[char][top  - 1])\n",
    "                bottom = self.count[char] + self.Occ[char][bottom] - 1\n",
    "            else:\n",
    "                return []\n",
    "\n",
    "            if top > bottom:\n",
    "                return []\n",
    "\n",
    "        return self.sa.suffix_array()[top:bottom + 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "bwt = BWTSearcher(genome_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 1/100:\t\t contain 0 mismatch\n",
      "Read 2/100:\t\t contain 0 mismatch\n",
      "Read 3/100:\t\t contain 0 mismatch\n",
      "Read 4/100:\t\t contain 0 mismatch\n",
      "Read 5/100:\t\t contain 0 mismatch\n",
      "Read 6/100:\t\t contain 0 mismatch\n",
      "Read 7/100:\t\t contain 0 mismatch\n",
      "Read 8/100:\t\t contain 0 mismatch\n",
      "Read 9/100:\t\t contain 0 mismatch\n",
      "Read 10/100:\t\t contain 0 mismatch\n",
      "Read 11/100:\t\t contain 0 mismatch\n",
      "Read 12/100:\t\t contain 0 mismatch\n",
      "Read 13/100:\t\t contain 0 mismatch\n",
      "Read 14/100:\t\t contain 0 mismatch\n",
      "Read 15/100:\t\t contain 0 mismatch\n",
      "Read 16/100:\t\t contain 0 mismatch\n",
      "Read 17/100:\t\t contain 0 mismatch\n",
      "Read 18/100:\t\t contain 0 mismatch\n",
      "Read 19/100:\t\t contain 0 mismatch\n",
      "Read 20/100:\t\t contain 0 mismatch\n",
      "Read 21/100:\t\t contain 0 mismatch\n",
      "Read 22/100:\t\t contain 0 mismatch\n",
      "Read 23/100:\t\t contain 0 mismatch\n",
      "Read 24/100:\t\t contain 0 mismatch\n",
      "Read 25/100:\t\t contain 0 mismatch\n",
      "Read 26/100:\t\t contain 0 mismatch\n",
      "Read 27/100:\t\t contain 0 mismatch\n",
      "Read 28/100:\t\t contain 0 mismatch\n",
      "Read 29/100:\t\t contain 0 mismatch\n",
      "Read 30/100:\t\t contain 0 mismatch\n",
      "Read 31/100:\t\t contain 0 mismatch\n",
      "Read 32/100:\t\t contain 0 mismatch\n",
      "Read 33/100:\t\t contain 0 mismatch\n",
      "Read 34/100:\t\t contain 0 mismatch\n",
      "Read 35/100:\t\t contain 0 mismatch\n",
      "Read 36/100:\t\t contain 0 mismatch\n",
      "Read 37/100:\t\t contain 0 mismatch\n",
      "Read 38/100:\t\t contain 0 mismatch\n",
      "Read 39/100:\t\t contain 0 mismatch\n",
      "Read 40/100:\t\t contain 0 mismatch\n",
      "Read 41/100:\t\t contain 0 mismatch\n",
      "Read 42/100:\t\t contain 0 mismatch\n",
      "Read 43/100:\t\t contain 0 mismatch\n",
      "Read 44/100:\t\t contain 0 mismatch\n",
      "Read 45/100:\t\t contain 0 mismatch\n",
      "Read 46/100:\t\t contain 0 mismatch\n",
      "Read 47/100:\t\t contain 0 mismatch\n",
      "Read 48/100:\t\t contain 0 mismatch\n",
      "Read 49/100:\t\t contain 0 mismatch\n",
      "Read 50/100:\t\t contain 0 mismatch\n",
      "Read 51/100:\t\t contain 0 mismatch\n",
      "Read 52/100:\t\t contain 0 mismatch\n",
      "Read 53/100:\t\t contain 0 mismatch\n",
      "Read 54/100:\t\t contain 0 mismatch\n",
      "Read 55/100:\t\t contain 0 mismatch\n",
      "Read 56/100:\t\t contain 0 mismatch\n",
      "Read 57/100:\t\t contain 0 mismatch\n",
      "Read 58/100:\t\t contain 0 mismatch\n",
      "Read 59/100:\t\t contain 0 mismatch\n",
      "Read 60/100:\t\t contain 0 mismatch\n",
      "Read 61/100:\t\t contain 0 mismatch\n",
      "Read 62/100:\t\t contain 0 mismatch\n",
      "Read 63/100:\t\t contain 0 mismatch\n",
      "Read 64/100:\t\t contain 0 mismatch\n",
      "Read 65/100:\t\t contain 0 mismatch\n",
      "Read 66/100:\t\t contain 0 mismatch\n",
      "Read 67/100:\t\t contain 0 mismatch\n",
      "Read 68/100:\t\t contain 0 mismatch\n",
      "Read 69/100:\t\t contain 0 mismatch\n",
      "Read 70/100:\t\t contain 0 mismatch\n",
      "Read 71/100:\t\t contain 0 mismatch\n",
      "Read 72/100:\t\t contain 0 mismatch\n",
      "Read 73/100:\t\t contain 0 mismatch\n",
      "Read 74/100:\t\t contain 0 mismatch\n",
      "Read 75/100:\t\t contain 0 mismatch\n",
      "Read 76/100:\t\t contain 0 mismatch\n",
      "Read 77/100:\t\t contain 0 mismatch\n",
      "Read 78/100:\t\t contain 0 mismatch\n",
      "Read 79/100:\t\t contain 0 mismatch\n",
      "Read 80/100:\t\t contain 0 mismatch\n",
      "Read 81/100:\t\t contain 0 mismatch\n",
      "Read 82/100:\t\t contain 0 mismatch\n",
      "Read 83/100:\t\t contain 0 mismatch\n",
      "Read 84/100:\t\t contain 0 mismatch\n",
      "Read 85/100:\t\t contain 0 mismatch\n",
      "Read 86/100:\t\t contain 0 mismatch\n",
      "Read 87/100:\t\t contain 0 mismatch\n",
      "Read 88/100:\t\t contain 0 mismatch\n",
      "Read 89/100:\t\t contain 0 mismatch\n",
      "Read 90/100:\t\t contain 0 mismatch\n",
      "Read 91/100:\t\t contain 0 mismatch\n",
      "Read 92/100:\t\t contain 0 mismatch\n",
      "Read 93/100:\t\t contain 0 mismatch\n",
      "Read 94/100:\t\t contain 0 mismatch\n",
      "Read 95/100:\t\t contain 0 mismatch\n",
      "Read 96/100:\t\t contain 0 mismatch\n",
      "Read 97/100:\t\t contain 0 mismatch\n",
      "Read 98/100:\t\t contain 0 mismatch\n",
      "Read 99/100:\t\t contain 0 mismatch\n",
      "Read 100/100:\t\t contain 0 mismatch\n"
     ]
    }
   ],
   "source": [
    "error_mutations_free_count, one_mutations_count, five_mutations_count = 0, 0, 0\n",
    "list_mutations_free, list_one_mutations, list_five_mutations = [], [], []\n",
    "read_counter = 0\n",
    "\n",
    "with open(\"BWT_folder/sample_reads.fasta\", \"r\") as file:\n",
    "    for record in SeqIO.parse(file, \"fasta\"):\n",
    "        read_sequence = str(record.seq)\n",
    "        read_counter += 1\n",
    "\n",
    "        occurrences = bwt.bwt_pattern_search(read_sequence)\n",
    "        num_occurrences = len(occurrences)\n",
    "\n",
    "        # print(occurrences)\n",
    "        \n",
    "        read_type = 0\n",
    "        if num_occurrences <= 1:\n",
    "            error_mutations_free_count += 1\n",
    "            list_mutations_free.append(read_counter)\n",
    "        \n",
    "        elif num_occurrences == 2:\n",
    "            read_type = 1\n",
    "            one_mutation_count += 1\n",
    "            list_one_mutations.append(read_counter)\n",
    "            \n",
    "        elif num_occurrences >= 6:\n",
    "            read_type = 5\n",
    "            five_mutations_count += 1\n",
    "            list_five_mutations.append(read_counter)\n",
    "\n",
    "        print(f\"Read {read_counter}/100:\\t\\t contain {read_type} mismatch\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "\n",
    "Это домашнее задание можно выполнять целиком в этом ноутбуке, либо алгоритмы написать в отдельном файле и импортировать сюда, для использования. В папке data лежат два файла islands.fasta и nonIslands.fasta. В них хранятся прочтения из CpG островков и из обычных участков генома соответственно, этими данными нужно будет воспользоваться в первом задании."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Задача №1 (1)\n",
    "Определите частоты генерации для каждого из нуклеотидов внутри CpG островков и вне их. Посчитайте так-же частоты для всех упорядоченных пар нуклеотидов и сравните частоту пары CG внутри островков и снаружи. Сделайте вывод. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in SeqIO.parse(\"data/islands.fasta\", \"fasta\"):\n",
    "    islands = str(record.seq)\n",
    "\n",
    "for record in SeqIO.parse(\"data/nonIslands.fasta\", \"fasta\"):\n",
    "    nonIslands = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t*** islands ***\n",
      "Частоты нуклеотидов: {'A': 0.1755, 'C': 0.3185, 'G': 0.3405, 'T': 0.1655}\n",
      "Частота CpG островков: 0.260596546310832\n",
      "Частота пары нуклеотидов CG: 0.0830415207603802\n",
      "\n",
      "\n",
      "\t*** nonIslands ***\n",
      "Частоты нуклеотидов: {'A': 0.28929861153616176, 'C': 0.19613640750872205, 'G': 0.1984481597792771, 'T': 0.3161168211758391}\n",
      "Частота CpG островков: 0.026909597196216814\n",
      "Частота пар нуклеотидов CG: 0.0052779765716222836\n"
     ]
    }
   ],
   "source": [
    "def count_nucleotide_frequency(sequence):\n",
    "    nucleotide_count = {'A': 0, 'C': 0, 'G': 0, 'T': 0}\n",
    "    total_nucleotides = len(sequence)\n",
    "    for nucl in sequence:\n",
    "        nucleotide_count[nucl] += 1\n",
    "    frequency = {nucl: count / total_nucleotides for nucl, count in nucleotide_count.items()}\n",
    "    return frequency\n",
    "\n",
    "def count_cpg_frequency(sequence):\n",
    "    cpg_count = sequence.count('CG')\n",
    "    c_count = sequence.count('C')\n",
    "    cpg_frequency = cpg_count / c_count if c_count > 0 else 0\n",
    "    return cpg_frequency\n",
    "\n",
    "def count_cg_pair_frequency(sequence):\n",
    "    pair_count = sequence.count('CG')\n",
    "    total_pairs = len(sequence) - 1\n",
    "    pair_frequency = pair_count / total_pairs if total_pairs > 0 else 0\n",
    "    return pair_frequency\n",
    "\n",
    "\n",
    "\n",
    "print(\"\\t*** islands ***\")\n",
    "nucleotide_frequency = count_nucleotide_frequency(islands)\n",
    "print(\"Частоты встречаемости нуклеотидов:\", nucleotide_frequency)\n",
    "\n",
    "cpg_frequency = count_cpg_frequency(islands)\n",
    "print(\"Частота CpG островков:\", cpg_frequency)\n",
    "\n",
    "cg_pair_frequency = count_cg_pair_frequency(islands)\n",
    "print(\"Частота пар нуклеотидов CG:\", cg_pair_frequency)\n",
    "\n",
    "\n",
    "print(\"\\n\\n\\t*** nonIslands ***\")\n",
    "nucleotide_frequency = count_nucleotide_frequency(nonIslands)\n",
    "print(\"Частоты нуклеотидов:\", nucleotide_frequency)\n",
    "\n",
    "cpg_frequency = count_cpg_frequency(nonIslands)\n",
    "print(\"Частота CpG островков:\", cpg_frequency)\n",
    "\n",
    "cg_pair_frequency = count_cg_pair_frequency(nonIslands)\n",
    "print(\"Частота пар нуклеотидов CG:\", cg_pair_frequency)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "212393\n",
      "2000\n"
     ]
    }
   ],
   "source": [
    "# \n",
    "print(len(nonIslands))\n",
    "print(len(islands))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Задача №2 (2)\n",
    "Напишите марковскую модель, которая имеет открытые состояния {A, T, G, C}, и скрытые состояния {+, -}. Когда модель в состоянии **+**, то вероятность генерации некоторого символа нуклеотида соответствует его частоте внутри CpG островков, вычислиному в первом задании, если состояние **-**, то частоте вне островков. Вероятность остаться внутри островка 0.95, а перейти в обычный геном 0.05. Для остальной части генома соответствующие вероятности 0.995 и 0.005. Саму модель можно реализовать в виде итератора, определив метод next, который возвращает пару - состояние и нуклеотид, который в этом состоянии произведен.    \n",
    "Воспользуйтесь данной моделью для того чтобы сгенерировать набор из 20 последовательностей длинной от 1 000 до 100 000, причем к каждой последовательности должна прилагаться последовательность состояний."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "class MarkovModel:\n",
    "    def __init__(self, nucleotide_freq_inside, nucleotide_freq_outside):\n",
    "        self.CpG_freq = nucleotide_freq_inside\n",
    "        self.non_CpG_freq = nucleotide_freq_outside\n",
    "        self.transition_prob_inside = 0.95\n",
    "        self.transition_prob_outside = 0.005\n",
    "        self.current_state = '+'\n",
    "        \n",
    "    def next(self):\n",
    "        if self.current_state == '+':\n",
    "            self.current_state = '-' if random.random() > self.transition_prob_inside else '+'\n",
    "            nucleotide = random.choices(['A', 'T', 'G', 'C'], weights=list(self.CpG_freq.values()))[0]\n",
    "        else:\n",
    "            self.current_state = '+' if random.random() > self.transition_prob_outside else '-'\n",
    "            nucleotide = random.choices(['A', 'T', 'G', 'C'], weights=list(self.non_CpG_freq.values()))[0]\n",
    "        return self.current_state, nucleotide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "CpG_freq = count_nucleotide_frequency(islands)\n",
    "non_CpG_freq = count_nucleotide_frequency(nonIslands)\n",
    "\n",
    "model = MarkovModel(CpG_freq, non_CpG_freq)\n",
    "\n",
    "sequences, states = [], []\n",
    "\n",
    "for _ in range(20):\n",
    "    sequence_length = random.randint(1000, 100000)\n",
    "    sequence = \"\"\n",
    "    state_sequence = \"\"\n",
    "    for _ in range(sequence_length):\n",
    "        state, nucleotide = model.next()\n",
    "        sequence += nucleotide\n",
    "        state_sequence += state\n",
    "    sequences.append(sequence)\n",
    "    states.append(state_sequence)\n",
    "\n",
    "\n",
    "with open(\"Task_2.log\", \"w\") as log:\n",
    "    for i, (seq, state) in enumerate(zip(sequences, states), 1):\n",
    "        log.write(f\"\\t\\t*** Sequence {i}/20  (length is {len(seq)}) ***\\n\\n\")\n",
    "        log.write(seq)\n",
    "        log.write(\"\\n\\nStates:\\n\")\n",
    "        log.write(state)\n",
    "        log.write(\"\\n\"*4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Задача №3 (4)\n",
    "Напишите алгоритм Витерби для восстановления последовательности скрытых состояний марковской модели из второго задаания. Воспользуйтесь им, воссстановив состояния тех последовательностей, которые вы получили во втором задании и посчитайте TP, TN, FP, FN по количеству правильно или ошибочно предсказанных позиций из CpG остравков. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### плохая реализация: TN == FN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Viterbi(MarkovModel):\n",
    "    def __init__(self, CpG_freq, non_CpG_freq):\n",
    "        self.CpG_freq = CpG_freq\n",
    "        self.non_CpG_freq = non_CpG_freq\n",
    "        self.current_state = '+'\n",
    "\n",
    "    \n",
    "    def viterbi(self, sequence):\n",
    "        T = len(sequence)\n",
    "        V = [{}]\n",
    "        path = {}\n",
    "\n",
    "        for state in ['+', '-']:\n",
    "            V[0][state] = 1.0\n",
    "            path[state] = [state]\n",
    "\n",
    "        for t in range(1, T):\n",
    "            V.append({})\n",
    "            new_path = {}\n",
    "\n",
    "            for next_state in ['+', '-']:\n",
    "                (prob, prev_state) = max(\n",
    "                    [(V[t - 1][prev_state] * self.transition_probability(prev_state, next_state) *\n",
    "                      self.emission_probability(next_state, sequence[t]), prev_state)\n",
    "                     for prev_state in ['+', '-']]\n",
    "                )\n",
    "                V[t][next_state] = prob\n",
    "                new_path[next_state] = path[prev_state] + [next_state]\n",
    "\n",
    "            path = new_path\n",
    "\n",
    "        (prob, state) = max((V[T - 1][final_state], final_state) for final_state in ['+', '-'])\n",
    "        return path[state]\n",
    "\n",
    "    def transition_probability(self, prev_state, next_state):\n",
    "        if prev_state == '+' and next_state == '+':\n",
    "            return 0.95\n",
    "        elif prev_state == '+' and next_state == '-':\n",
    "            return 0.05\n",
    "        elif prev_state == '-' and next_state == '-':\n",
    "            return 0.995\n",
    "        elif prev_state == '-' and next_state == '+':\n",
    "            return 0.005\n",
    "\n",
    "    def emission_probability(self, state, nucleotide):\n",
    "        if state == '+':\n",
    "            return self.CpG_freq[nucleotide]\n",
    "        else:\n",
    "            return self.non_CpG_freq[nucleotide]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*** Sequence 1/20  (length is 17285) ***\n",
      "\n",
      "TP: 0\tTN: 772\tFP: 0\tFN: 772\n",
      "\n",
      "\n",
      "*** Sequence 2/20  (length is 58297) ***\n",
      "\n",
      "TP: 23\tTN: 2797\tFP: 2\tFN: 2797\n",
      "\n",
      "\n",
      "*** Sequence 3/20  (length is 54202) ***\n",
      "\n",
      "TP: 86\tTN: 2551\tFP: 6\tFN: 2551\n",
      "\n",
      "\n",
      "*** Sequence 4/20  (length is 23387) ***\n",
      "\n",
      "TP: 0\tTN: 1161\tFP: 0\tFN: 1161\n",
      "\n",
      "\n",
      "*** Sequence 5/20  (length is 10466) ***\n",
      "\n",
      "TP: 140\tTN: 474\tFP: 2\tFN: 474\n",
      "\n",
      "\n",
      "*** Sequence 6/20  (length is 50368) ***\n",
      "\n",
      "TP: 0\tTN: 2451\tFP: 0\tFN: 2451\n",
      "\n",
      "\n",
      "*** Sequence 7/20  (length is 24254) ***\n",
      "\n",
      "TP: 0\tTN: 1189\tFP: 0\tFN: 1189\n",
      "\n",
      "\n",
      "*** Sequence 8/20  (length is 54706) ***\n",
      "\n",
      "TP: 0\tTN: 2575\tFP: 0\tFN: 2575\n",
      "\n",
      "\n",
      "*** Sequence 9/20  (length is 41706) ***\n",
      "\n",
      "TP: 65\tTN: 1989\tFP: 4\tFN: 1989\n",
      "\n",
      "\n",
      "*** Sequence 10/20  (length is 65818) ***\n",
      "\n",
      "TP: 99\tTN: 3149\tFP: 4\tFN: 3149\n",
      "\n",
      "\n",
      "*** Sequence 11/20  (length is 67363) ***\n",
      "\n",
      "TP: 81\tTN: 3226\tFP: 2\tFN: 3226\n",
      "\n",
      "\n",
      "*** Sequence 12/20  (length is 58150) ***\n",
      "\n",
      "TP: 0\tTN: 2820\tFP: 0\tFN: 2820\n",
      "\n",
      "\n",
      "*** Sequence 13/20  (length is 26101) ***\n",
      "\n",
      "TP: 0\tTN: 1242\tFP: 0\tFN: 1242\n",
      "\n",
      "\n",
      "*** Sequence 14/20  (length is 74449) ***\n",
      "\n",
      "TP: 0\tTN: 3573\tFP: 0\tFN: 3573\n",
      "\n",
      "\n",
      "*** Sequence 15/20  (length is 50242) ***\n",
      "\n",
      "TP: 49\tTN: 2361\tFP: 1\tFN: 2361\n",
      "\n",
      "\n",
      "*** Sequence 16/20  (length is 40119) ***\n",
      "\n",
      "TP: 16\tTN: 1921\tFP: 0\tFN: 1921\n",
      "\n",
      "\n",
      "*** Sequence 17/20  (length is 86885) ***\n",
      "\n",
      "TP: 0\tTN: 4129\tFP: 0\tFN: 4129\n",
      "\n",
      "\n",
      "*** Sequence 18/20  (length is 81630) ***\n",
      "\n",
      "TP: 0\tTN: 3950\tFP: 0\tFN: 3950\n",
      "\n",
      "\n",
      "*** Sequence 19/20  (length is 79366) ***\n",
      "\n",
      "TP: 0\tTN: 3940\tFP: 0\tFN: 3940\n",
      "\n",
      "\n",
      "*** Sequence 20/20  (length is 39508) ***\n",
      "\n",
      "TP: 71\tTN: 1999\tFP: 3\tFN: 1999\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def calculate_metrics(true_states, predicted_states):\n",
    "    TP, TN, FP, FN = 0, 0, 0, 0\n",
    "    for true_state, pred_state in zip(true_states, predicted_states):\n",
    "        if true_state == '+' and pred_state == '+':\n",
    "            TP += 1\n",
    "        elif true_state == '-' and pred_state == '-':\n",
    "            TN += 1\n",
    "        elif true_state == '-' and pred_state == '+':\n",
    "            FP += 1\n",
    "        elif true_state == '+' and pred_state == '-':\n",
    "            FN += 1\n",
    "    return TP, TN, FP, FN\n",
    "\n",
    "\n",
    "model_viterbi = Viterbi(CpG_freq, non_CpG_freq)\n",
    "for i, (sequence, true_states) in enumerate(zip(sequences, states), 1):\n",
    "    predicted_states = model_viterbi.viterbi(sequence)\n",
    "    tp, tn, fp, fn = calculate_metrics(true_states, predicted_states)\n",
    "\n",
    "    print(f\"*** Sequence {i}/20  (length is {len(sequence)}) ***\\n\")\n",
    "    print(f\"TP: {tp}\\tTN: {tn}\\tFP: {fp}\\tFN: {tn}\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Задача №4 (4)\n",
    "Напишите алгоритм вперед назад для модели из второго задания. Пользуясь этим алгоритмом найдите вероятности того, что модель находилась в состоянии **+** для каждой позиции строк из второго задания. Устанавливая различные пороговые значения, определите позиции соответствующие CpG островкам и посчитайте TP. Постройте график зависимости TP от выбранного порогового значения. Есть ли пороговые значения при которых TP больше чем в задании №3?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ForwardBackward(MarkovModel):\n",
    "    def forward_backward(self, sequence):\n",
    "        T = len(sequence)\n",
    "        forward = [{} for _ in range(T)]\n",
    "        backward = [{} for _ in range(T)]\n",
    "        \n",
    "        # forward\n",
    "        for t in range(T):\n",
    "            if t == 0:\n",
    "                for state in ['+', '-']:\n",
    "                    forward[t][state] = 1.0\n",
    "            else:\n",
    "                for next_state in ['+', '-']:\n",
    "                    forward[t][next_state] = sum(\n",
    "                        forward[t - 1][state] * self.transition_probability(state, next_state) * \n",
    "                        self.emission_probability(next_state, sequence[t])\n",
    "                        for state in ['+', '-']\n",
    "                    )\n",
    "\n",
    "            # возникала ошибка при делении на 0\n",
    "            total_forward_prob = sum(forward[t].values())\n",
    "            for state in forward[t]:\n",
    "                forward[t][state] /= total_forward_prob if total_forward_prob != 0 else 1.0\n",
    "        \n",
    "\n",
    "        \n",
    "        # backward\n",
    "        for t in range(T - 1, -1, -1):\n",
    "            if t == T - 1:\n",
    "                for state in ['+', '-']:\n",
    "                    backward[t][state] = 1.0\n",
    "            else:\n",
    "                for prev_state in ['+', '-']:\n",
    "                    backward[t][prev_state] = sum(\n",
    "                        self.transition_probability(prev_state, next_state) * \n",
    "                        self.emission_probability(next_state, sequence[t + 1]) * \n",
    "                        backward[t + 1][next_state]\n",
    "                        for next_state in ['+', '-']\n",
    "                    )\n",
    "\n",
    "            total_backward_prob = sum(backward[t].values())\n",
    "            for state in backward[t]:\n",
    "                backward[t][state] /= total_backward_prob if total_backward_prob != 0 else 1.0\n",
    "        \n",
    "        posterior_probs = []\n",
    "        for t in range(T):\n",
    "            posterior_probs.append({\n",
    "                '+': forward[t]['+'] * backward[t]['+'],\n",
    "                '-': forward[t]['-'] * backward[t]['-']\n",
    "            })\n",
    "            \n",
    "        return posterior_probs\n",
    "    \n",
    "    def transition_probability(self, prev_state, next_state):\n",
    "        if prev_state == '+' and next_state == '+':\n",
    "            return self.transition_prob_inside\n",
    "        elif prev_state == '+' and next_state == '-':\n",
    "            return 1 - self.transition_prob_inside\n",
    "        elif prev_state == '-' and next_state == '-':\n",
    "            return 1 - self.transition_prob_outside\n",
    "        elif prev_state == '-' and next_state == '+':\n",
    "            return self.transition_prob_outside\n",
    "\n",
    "    def emission_probability(self, state, nucleotide):\n",
    "        if state == '+':\n",
    "            return self.CpG_freq[nucleotide]\n",
    "        else:\n",
    "            return self.non_CpG_freq[nucleotide]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'TP_by_sequence' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 22\u001b[0m\n\u001b[1;32m     18\u001b[0m         TP_by_seq\u001b[38;5;241m.\u001b[39mappend(TP)\n\u001b[1;32m     20\u001b[0m     TP_by_seq\u001b[38;5;241m.\u001b[39mappend(TP_by_seq)\n\u001b[0;32m---> 22\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, TP_by_seq \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[43mTP_by_sequence\u001b[49m, \u001b[38;5;241m1\u001b[39m):\n\u001b[1;32m     23\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m*** Sequence \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/20  (length is \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(sequence)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m) ***\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m threshold, TP \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(thresholds, TP_by_seq):\n",
      "\u001b[0;31mNameError\u001b[0m: name 'TP_by_sequence' is not defined"
     ]
    }
   ],
   "source": [
    "model_forward_backward = ForwardBackward(CpG_freq, non_CpG_freq)\n",
    "\n",
    "posterior_probs_list = []\n",
    "for sequence in sequences:\n",
    "    posterior_probs_list.append(model_forward_backward.forward_backward(sequence))\n",
    "\n",
    "thresholds = [0.5, 0.6, 0.7, 0.8, 0.9]\n",
    "\n",
    "for sequence, true_state in zip(sequences, states):\n",
    "    \n",
    "    posterior_probs = model_forward_backward.forward_backward(sequence)\n",
    "    \n",
    "    TP_by_seq = []\n",
    "    \n",
    "    for threshold in thresholds:\n",
    "        predicted_states = ['+' if posterior_prob['+'] > threshold else '-' for posterior_prob in posterior_probs]\n",
    "        TP = sum(1 for pred_state, state in zip(predicted_states, true_state) if pred_state == '+' and true_state == '+')\n",
    "        TP_by_seq.append(TP)\n",
    "    \n",
    "    TP_by_seq.append(TP_by_seq)\n",
    "\n",
    "for i, TP_by_seq in enumerate(TP_by_seq, 1):\n",
    "    print(f\"*** Sequence {i}/20  (length is {len(sequence)}) ***\\n\")\n",
    "    for threshold, TP in zip(thresholds, TP_by_seq):\n",
    "        print(f\"Threshold: {threshold}\\t TP: {TP}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
